File: algorithm_2019/univalued-binary-tree.py
"""
如果二叉树每个节点都具有相同的值，那么该二叉树就是单值二叉树。

只有给定的树是单值二叉树时，才返回 true；否则返回 false。



示例 1：



输入：[1,1,1,1,1,null,1]
输出：true
示例 2：



输入：[2,2,2,5,2]
输出：false


提示：

给定树的节点数范围是 [1, 100]。
每个节点的值都是整数，范围为 [0, 99] 。
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isUnivalTree(self, root) -> bool:
        nums = set()
        stack = [root]
        while stack:
            node = stack.pop()
            if nums and node.val not in nums:
                return False
            nums.add(node.val)
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)
        return True

File: Ruby/univalued-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

def traverse(node, values)
  if node.nil?
    return true
  end
  if !values.include?(node.val)
    return false
  end
  values.push(node.val)
  if !traverse(node.left, values)
    return false
  end
  if !traverse(node.right, values)
    return false
  end
  return true
end


# @param {TreeNode} root
# @return {Boolean}
def is_unival_tree(root)
  values = []
  values.push(root.val)
  traverse(root, values)
end
